web-platform/website/src/services/AttributionService.tsx

```diff
@@ -0,0 +1,59 @@
+import { createContext, useContext, useEffect, useState } from "react";
+
+export interface AttributionDataContext {
+  // Attribution codes from local storage.
+  allAttributionCodes: URLSearchParams;
+
+  // Loading flag for all attribution info.
+  isLoaded: boolean;
+}
+
+const AttributionContext = createContext<AttributionDataContext>(
+  {} as AttributionDataContext
+);
+
+export function useAttribution() {
+  return useContext(AttributionContext);
+}
+
+export function AttributionProvider(props: { children: React.ReactNode }) {
+  const [isLoaded, setIsLoaded] = useState(false);
+  const [allAttributionCodes, setAllAttributionCodes] = useState<URLSearchParams>(new URLSearchParams());
+
+  async function loadAllAttributionCodes() {
+    // Add UTM codes to store links, if present.
+    let utmParams = new URLSearchParams();
+    if (typeof window !== "undefined") {
+      const utmKeys = ["source", "medium", "campaign", "term", "content"];
+
+      // Use "last-click" attribution, overwriting any stored UTM codes if new ones are encountered.
+      const storedParams = window.localStorage.getItem("rally_utm");
+      if (storedParams) {
+        utmParams = new URLSearchParams(storedParams);
+      }
+
+      const searchParams = (new URL(document.location.href)).searchParams;
+      utmKeys.forEach((key) => {
+        const utmKey = `utm_${key}`;
+        searchParams.has(utmKey) && utmParams.set(utmKey, searchParams.get(utmKey) ?? "");
+      });
+      window.localStorage.setItem("rally_utm", utmParams.toString());
+
+    }
+    setAllAttributionCodes(utmParams);
+
+    setIsLoaded(true);
+  }
+
```

Comment:
We can add another method here setAttributionCode(url) and bring in the logic of merging UTM codes with query params (if any) and return the resulting URL. Then from StudiesService we can just call setAttributionCode.